File: 39_CombinationSum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        total_set = []
        def helper(cands, target, curr=[]):
            if target < 0 or len(cands) <= 0:
                return total_set
            elif target == 0:
                total_set.append(curr)
                return total_set
            else:
                helper(cands, target-cands[0], curr + [cands[0]])
                helper(cands[1:], target, curr)
            return total_set
        return helper(candidates, target, [])   

File: kotlin/39_CombinationSum.kt
class Solution {
    var results: MutableList<MutableList<Int>> = MutableList(0){MutableList(0){0}}
    fun combinationSum(candidates: IntArray, target: Int): List<List<Int>> {
        fun combine(idx: Int, sum: Int, currAttempt: MutableList<Int>) {
            if (sum == target) {
                results.add(currAttempt.toMutableList())
                return
            } else if (sum > target) {
                return
            }
            for (i in idx until candidates.size) {
                var temp = MutableList(currAttempt.size){it->currAttempt[it]}
                temp.add(candidates[i])
                combine(i, sum + candidates[i], temp)
            }
        }
        combine(0, 0, mutableListOf())
        return results
    }
}
